base/networking.c

```diff
@@ -762,3 +762,340 @@ ipv6_is_enabled ()
+        {
+          /* addr is 0.0.0.0 */
+          if (addr6_p->s6_addr32[3] == 0)
+            return 1;
+
+          /* addr starts with 127.0.0.1 */
+          if ((addr6_p->s6_addr32[3] & htonl (0xFF000000))
+              == htonl (0x7F000000))
+            return 1;
+        }
+      if (IN6_IS_ADDR_LOOPBACK (addr6_p))
+        return 1;
+    }
+
+  if (getifaddrs (&ifaddr) == -1)
+    {
+      g_debug ("%s: getifaddr failed: %s", __func__, strerror (errno));
+      return FALSE;
+    }
+  else
+    {
+      struct sockaddr_in *sin;
+      struct sockaddr_in6 *sin6;
+
+      for (ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next)
+        {
+          if (ifa->ifa_addr == NULL)
+            continue;
+          if (ifa->ifa_addr->sa_family == AF_INET)
+            {
+              sin = (struct sockaddr_in *) (ifa->ifa_addr);
+              /* Check if same address as local interface. */
+              if (addr_p->s_addr == sin->sin_addr.s_addr)
+                return TRUE;
+            }
+          if (ifa->ifa_addr->sa_family == AF_INET6)
+            {
+              sin6 = (struct sockaddr_in6 *) (ifa->ifa_addr);
+              /* Check if same address as local interface. */
+              if (IN6_ARE_ADDR_EQUAL (&(sin6->sin6_addr), addr6_p))
+                return TRUE;
+            }
+        }
+      freeifaddrs (ifaddr);
+    }
+
+  return FALSE;
+}
+
+typedef struct route_entry route_entry_t;
+
+/** Entry of routing table /proc/net/route */
+struct route_entry
+{
+  gchar *interface;
+  unsigned long mask;
+  unsigned long dest;
+};
+
+/**
+ * @brief Get the entries of /proc/net/route as list of route_entry structs.
+ *
+ * @return  GSList of route_entry structs.
+ */
+static GSList *
+get_routes (void)
+{
+  GSList *routes = NULL;
+  GError *err = NULL;
+  GIOChannel *file_channel;
+  gchar *line;
+  gchar **items_in_line;
+  int status;
+  route_entry_t *entry;
+
+  /* Open "/proc/net/route". */
+  file_channel = g_io_channel_new_file ("/proc/net/route", "r", &err);
+  if (file_channel == NULL)
+    {
+      g_warning ("%s: %s. ", __func__,
+                 err ? err->message : "Error opening /proc/net/ipv6_route");
+      err = NULL;
+      return NULL;
+    }
+
+  /* Skip first first line of file. */
+  status = g_io_channel_read_line (file_channel, &line, NULL, NULL, &err);
+  if (status != G_IO_STATUS_NORMAL || !line || err)
+    {
+      g_warning ("%s: %s", __func__,
+                 err ? err->message
+                     : "g_io_channel_read_line() status != G_IO_STATUS_NORMAL");
+      if (line)
+        g_free (line);
+      err = NULL;
+    }
+
+  /* Until EOF or err we go through lines of file and extract Iface, Mask and
+   * Destination and put it into the to be returned list of routes.*/
+  while (1)
+    {
+      gchar *interface, *char_p;
+      unsigned long mask, dest;
+      int count = 0;
+
+      /* Get new line. */
+      status = g_io_channel_read_line (file_channel, &line, NULL, NULL, &err);
```

Comment:
This still looks like a leak to me.  A line is allocated by the g_io_channel_read_line called before the while, but that line is not freed before the call to g_io_channel_read_line here.